### PR TITLE
Set SpanContext on SentryTransaction to avoid potential NPE.

### PR DESCRIPTION
**Pull request by maciejwalkowiak**
## :scroll: Description

Set trace context on SentryTransaction and pass it to `contexts` on `transaction#finish`.

## :bulb: Motivation and Context

The alternative to https://github.com/getsentry/sentry-java/pull/1164

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

**Comment by maciejwalkowiak** [comment]
@bruno-garcia @marandaneto I think this is a better than #1164 approach - the result the discussion we had. I think trace context on `finish` is cleaner than doing it during serialization.

**Review by bruno-garcia — APPROVED** [review]
Much nicer!

File: sentry/src/main/java/io/sentry/SentryTransaction.java [hunk 1]
```diff
@@ -57,7 +59,7 @@ public SentryTransaction(
     this.transaction = Objects.requireNonNull(name, "name is required");
     this.startTimestamp = DateUtils.getCurrentDateTime();
     this.hub = Objects.requireNonNull(hub, "hub is required");
-    this.getContexts().setTrace(contexts);
+    this.context = contexts;
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/SentryTransaction.java:62]: do the assignment with `Objects.requireNonNull(contexts, "contexts are required");`

File: sentry/src/main/java/io/sentry/SentryTransaction.java [hunk 2]
```diff
@@ -28,7 +28,9 @@
    * A hub this transaction is attached to. Marked as transient to be ignored during JSON
    * serialization.
    */
-  private @NotNull final transient IHub hub;
+  private final transient @NotNull IHub hub;
+
+  private final transient @NotNull SpanContext context;
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/SentryTransaction.java:33]:
```suggestion
  private final transient @NotNull SpanContext contexts;
```

File: sentry/src/main/java/io/sentry/SentryTransaction.java [hunk 3]
```diff
@@ -239,4 +242,8 @@ Date getTimestamp() {
     }
     return null;
   }
+
+  SpanContext getContext() {
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/SentryTransaction.java:246]:
```suggestion
  @NotNull SpanContext getContext() {
```

File: sentry/src/main/java/io/sentry/SentryTransaction.java (hunk 2 above)
Comment (bruno-garcia) [reply sentry/src/main/java/io/sentry/SentryTransaction.java:33]: Wasn't the goal to move the name to `context`? Since `SpanContext`?

Comment (marandaneto) [reply sentry/src/main/java/io/sentry/SentryTransaction.java:33]: no goal, but I'd rather prefer the keep consistency, either the ctor param or the global field should rename, so they match.
it's a nitpick but why not naming the same?

File: sentry/src/main/java/io/sentry/SentryTransaction.java [hunk 4]
```diff
@@ -156,6 +158,7 @@ public void finish() {
     if (this.throwable != null) {
       hub.setSpanContext(this.throwable, this.getSpanContext());
     }
+    this.getContexts().setTrace(this.context);
```
Comment (marandaneto) [inline sentry/src/main/java/io/sentry/SentryTransaction.java:161]: here `this.getContexts()` is guaranteed to not be null, right?

Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/SentryTransaction.java:161]: Yes.

File: sentry/src/main/java/io/sentry/SentryTransaction.java (hunk 2 above)
Comment (maciejwalkowiak) [reply sentry/src/main/java/io/sentry/SentryTransaction.java:33]: renamed to "context"

**Review by marandaneto — APPROVED** [review]

**Pull request merged**